draft-ietf-quic-transport.md

```diff
@@ -2251,16 +2253,19 @@ transport parameter in the TLS handshake.
 Servers MAY communicate a preferred address of each address family (IPv4 and
 IPv6) to allow clients to pick the one most suited to their network attachment.
 
-Once the handshake is finished, the client SHOULD select one of the two
-server's preferred addresses and initiate path validation (see
-{{migrate-validate}}) of that address using the connection ID provided in the
-preferred_address transport parameter.
+Once the handshake is confirmed (see Section 4.1.2 of {{QUIC-TLS}}), the client
+SHOULD select one of the two server's preferred addresses and initiate path
+validation (see {{migrate-validate}}) of that address using an active unused
+connection ID.
 
 If path validation succeeds, the client SHOULD immediately begin sending all
-future packets to the new server address using the new connection ID and
-discontinue use of the old server address.  If path validation fails, the client
-MUST continue sending all future packets to the server's original IP address.
-
+future packets to the new server and discontinue use of the old server address,
+and retire the connection ID that was used for the original path
+({{retiring-cids}}).  If path validation fails, or if the client decides not to
+migrate to the preferred address, it MUST continue sending all future packets to
+the server's original IP address, and retire the connection ID provided by the
+preferred_address transport parameter.  Retirement of either of these connection
+IDs notifies the server of the address the client has chosen.
```

Comment:
Let me try to be clearer about my objection to this last sentence.

This implies a semantic to the retirement of connection IDs that is not already defined.  It says that in addition to releasing the resource, the server can say definitively that those other network paths won't be used.  But this is misleading because retiring CID 1 does not prevent CID 4 from being used on that path.  Nothing says that the connection IDs sent in NEW_CONNECTION_ID have to be used on one or other path.  Better to keep the requirement where it is: don't migrate back if you use a preferred address.  

Yes, that means that servers can't be sure of behaviour of clients here, but they can use the destination address to confirm acceptance of the preferred address or not.  That should suffice.

Reply:
Maybe you want to state that in the issue (i.e. #3353), as we've been discussing there about having this requirement?